core/src/main/java/org/kiwix/kiwixmobile/core/utils/files/FileLogger.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2020 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.kiwix.kiwixmobile.core.utils.files
+
+import android.content.Context
+import android.os.Environment
+import android.util.Log
+import java.io.File
+import java.io.IOException
+import java.lang.System.currentTimeMillis
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Authored by s-ayush2903 on 19 June 2020 */
+
+/** A class for writing logs to the file in the device */
+@Singleton
+class FileLogger @Inject constructor() {
+
+  /* Checks if external storage is available for read and write */
+  val isExternalStorageWritable: Boolean =
+    Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()
+
+  fun writeLogFile(context: Context): File {
+    // Create a new folder in private storage with name: logs
+    val logDir = File(context.filesDir, "logs")
+    val logFile = File(logDir, fileName)
+
+    Log.d("KIWIX", "Writing all logs into [" + logDir.path + "]")
+
+    if (!logDir.exists()) {
+      logDir.mkdir()
+    }
+    // Delete the pre-existing file
+    if (logDir.exists() && logFile.isFile) {
+      Log.d(TAG, "writeLogFile: Deleting preExistingFile")
+      logFile.delete()
+    }
+
+    try {
+      logDir.createNewFile()
+      Runtime.getRuntime().exec("logcat -c")
+      Runtime.getRuntime().exec("logcat -f $logFile -s kiwix")
+      Log.d(TAG, "writeLogFile: Log report written Successfully!")
+      Log.d(TAG, "writeLogFile: Authored report to $logFile")
+    } catch (e: IOException) {
+      Log.e("KIWIX", "Error while writing $fileName! ", e)
+    }
+    return logFile
+  }
+
+  companion object {
+    private const val TAG = "FileLogger"
+    private val fileName = "logs" + currentTimeMillis() + ".txt"
```

Comment:
this will always be the same value per app instance, you can use auto fixes and 
![image](https://user-images.githubusercontent.com/3358282/86892311-00703f00-c0f8-11ea-883d-5bdb2134d1f8.png)
or
![image](https://user-images.githubusercontent.com/3358282/86892369-1120b500-c0f8-11ea-85fa-213b6539de14.png)

Reply:
> this will always be the same value per app instance

* How can this be same for a particular instance, i mean doesn't it make use of the time at that particular instant while calling this method?
* Can you please help me let know one more thing that why getter initializer is being preferred over normal one, does it has some edge over the normal declaration?
* And one more question(just out of curiosity), i don't think that you had typed those changes in your machine and run, so how did you get to this, i mean did it just clicked you out of experience. (i was talking about the quick-fix suggestion provided)

Reply:
File Logger is marked `@Singleton` this means that every time you ask dagger for a FileLogger you get exactly the same one.

Companion objects are sort of how you declare "static" things in kotlin, all instances of a class share the same companion object. So even if you were getting new instances of fileLogger this value would be unchanged from the first time it is intialised

Properties in kotlin are more complex than simple fields in java because they do a lot more, [docs](https://kotlinlang.org/docs/reference/properties.html). By specifying a getter on a property every time you read from that property it calls the getter. It basically makes it identical to a method except you access it like `object.myproperty` vs `object.myProperty()`, you can check the coding convention [docs](https://kotlinlang.org/docs/reference/properties.html) to see when you should use which but preference enters it a bit.

For the quick fixes I wanted these precise quick fixes but couldn't remember the exact text so I thought pictures would be most specific. I found a property in my project, pressed `alt+enter` and it gave me the list of refactorings it will give for any property in the project and I took 2 screenshots and copy pasted here